File: zoho/L92-reverse_linked_list_2.py
# https://leetcode.com/problems/reverse-linked-list-ii/
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        prev, curr = dummy, head
        for i in range(left - 1):
            prev, curr = curr, curr.next
        p = None
        for i in range(right - left + 1):
            tempNext = curr.next
            curr.next = p
            p = curr
            curr = tempNext
        prev.next.next = curr
        prev.next = p 
        return dummy.next

File: zoho/L92-ReverseLinkedList2.java
// https://leetcode.com/problems/reverse-linked-list-ii/
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dummy = new ListNode(0, head);
        ListNode prev = dummy;
        ListNode curr = head;
        for(int i = 0; i < left - 1; i++) {
            prev = curr;
            curr = curr.next;
        }
        ListNode p = new ListNode();
        for(int i = 0; i < right - left + 1; i++) {
            ListNode tempNext = curr.next;
            curr.next = p;
            p = curr;
            curr = tempNext;
        }
        prev.next.next = curr;
        prev.next = p;
        return dummy.next;
    }
}
